text/0000-untyped-numeric-constants.md

```diff
@@ -0,0 +1,71 @@
+- Feature Name: untyped_numeric_constant
+- Start Date: 2017-04-06
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allows numeric constants to be defined without specifying an explicit type. Each time that constant is used, the type will be inferred as if it was an untyped numeric literal.
+
+# Motivation
+[motivation]: #motivation
+
+The major motivation for this is FFI bindings, in particular when converting a constant of the form `#define FOO 273`. The equivalent in Rust would be `pub const FOO: ??? = 273;`, but figuring out which type to use can often be difficult. In some cases it can even be impossible, where a given constant is used in multiple places each with a different type. In `winapi` alone there are over twenty five thousand constants, and those decisions make up a significant amount of time when writing the bindings and cannot be automated. Even worse, if the wrong decision is made, it cannot later be changed in the future without causing a breaking change. Untyped numeric constants would solve this completely.
+
+Even crates which aren't FFI bindings, but provide a variety of numerical constants can benefit from this feature.
+
+# Detailed design
+[design]: #detailed-design
+
+Allow untyped numeric constants to be defined like so:
+
+```Rust
+const INTEGER = 273;
+const FLOAT = 4.2;
+```
+
+When an untyped numeric constant is used, it will be equivalent to using the literal directly, and type inference will function identically. Thus the following two statements are identical in behavior (assuming `a` and `b` are later used identically) :
+
+```Rust
+let a = INTEGER + 5;
+let b = 273 + 5;
+```
+
+The type of an untyped numeric constant can be different each time it is used. Thus the following is valid:
+
+```Rust
+let a: i32 = INTEGER;
+let b: u32 = INTEGER;
+```
+
+No support for inferring integer constants as floats or vice versa is being proposed.
+
+# How We Teach This
+[how-we-teach-this]: #how-we-teach-this
+
+Just call them untyped integer constants or untyped float constants or untyped numeric constants.
+
+Teach them as an extension of standard constants.
+
+People who use such constants won't even notice the difference so there's no pressing need to teach it right away.
+
+It will require additions to all the documentation with regards to constants.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* New feature that has to be supported.
+* Can get in the way of a future RFC for inferred constants.
+
+# Alternatives
+[alternatives]: #alternatives
+
+* One alternative is to use macro constants, which involves defining a macro for each constant, and invoking the constant via `FOO!()` instead of `FOO`. It is verbose and ugly, clutters the global macro namespace, and will probably uncover some performance regression in rustc.
```

Comment:
@cuviper has suggested another alternative over at irlo: making it possible to drop the `()` at the macro invocation, requiring you only to type `FOO!`. Ofc still one character lost, and ultil macros 2.0 the global namespace problem still persists.